File: mobile/android/app/src/main/java/com/quickin/app/CurrencyManager.kt
package com.quickin.app

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.snapshots.SnapshotStateMap
import androidx.compose.runtime.mutableStateMapOf
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

/**
 * Multi-currency DISPLAY holder (Section 9 — MOCK). A process-wide singleton that:
 *  • holds the FX rates (EGP base + USD/EUR/GBP/SAR/AED), seeded with baked-in [STATIC_RATES] so the
 *    app formats correctly before — or without ever — hitting the network,
 *  • on [init] loads `GET /api/local/currencies` once and merges any returned rates over the
 *    baked-in ones (a network failure is silently ignored — the static rates remain),
 *  • persists the user's chosen display currency in SharedPreferences (key [KEY_CURRENCY], default
 *    "EGP") so it survives relaunch, and
 *  • exposes the current selection + rates as Compose state ([currency], [rates]) so every price
 *    [Text] that calls [format] recomposes the instant the user picks a new currency.
 *
 * Conversion is display-only: `amountInTarget = amountEgp * rates[target]`. All backend amounts —
 * and every booking / payment — stay in EGP regardless of what's shown.
 */
object CurrencyManager {

    /** The platform base currency. All amounts passed to [format] are in this currency. */
    const val BASE = "EGP"

    /** SharedPreferences file + key for the persisted display-currency choice. */
    private const val PREFS_NAME = "qk_currency_prefs"
    const val KEY_CURRENCY = "qk_currency"

    /**
     * The currencies the switcher offers, in display order. The first (EGP) is the base. Each entry
     * is the ISO code; [symbolFor] resolves a display symbol. Kept as the source of truth for the
     * picker even when the backend returns extra/fewer rates.
     */
    val SUPPORTED: List<String> = listOf("EGP", "USD", "EUR", "GBP", "SAR", "AED")

    /**
     * Baked-in fallback rates (EGP → target multiplier), matching the backend's
     * `GET /api/local/currencies` so display is correct offline / before the fetch resolves.
     */
    val STATIC_RATES: Map<String, Double> = linkedMapOf(
        "EGP" to 1.0,
        "USD" to 0.0203,
        "EUR" to 0.0188,
        "GBP" to 0.016,
        "SAR" to 0.0762,
        "AED" to 0.0746
    )

    /** Live FX rates as Compose state — seeded from [STATIC_RATES], overlaid by the fetch. */
    private val _rates: SnapshotStateMap<String, Double> =
        mutableStateMapOf<String, Double>().apply { putAll(STATIC_RATES) }
    val rates: Map<String, Double> get() = _rates

    /** The currently-selected display currency as Compose state (defaults to [BASE]). */
    private val currencyState = mutableStateOf(BASE)
    val currency: String get() = currencyState.value

    /** True once [init] has been called (so a second call from another screen is a no-op). */
    @Volatile
    private var initialized = false

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    /**
     * Loads the persisted currency choice and (once) refreshes the rates from the backend. Safe to
     * call from any composable's first appearance — repeat calls after the first are no-ops. On a
     * network failure the baked-in [STATIC_RATES] simply remain in effect.
     */
    fun init(context: Context) {
        if (initialized) return
        initialized = true
        val saved = prefs(context).getString(KEY_CURRENCY, BASE)
        currencyState.value = normalize(saved)
        scope.launch {
            val fetched = SupabaseService.fetchCurrencies()
            if (fetched != null && fetched.rates.isNotEmpty()) {
                // Merge over the baked-in rates so any currency the backend omits still formats.
                _rates.putAll(fetched.rates)
            }
        }
    }

    /**
     * Sets and persists the display currency (a no-op for an unknown code). Updates the Compose
     * [currency] state so price Text recomposes immediately.
     */
    fun setCurrency(context: Context, code: String) {
        val next = normalize(code)
        if (next == currencyState.value) return
        currencyState.value = next
        prefs(context).edit().putString(KEY_CURRENCY, next).apply()
    }

    /**
     * Formats an EGP [amountEgp] in the currently-selected display currency, e.g. "EGP 1,200" or
     * "$24" or "€22". Converts via the live [rates] (falling back to 1.0 for an unknown code, i.e.
     * the original EGP magnitude). Reading [currency] / [rates] here makes any composable that calls
     * this recompose when the selection or rates change.
     */
    fun format(amountEgp: Double): String {
        val code = currencyState.value
        val rate = _rates[code] ?: 1.0
        val converted = amountEgp * rate
        return symbolFor(code) + formatAmount(converted, code)
    }

    /** Convenience overload for Int EGP amounts. */
    fun format(amountEgp: Int): String = format(amountEgp.toDouble())

    /**
     * A display symbol/prefix for [code], e.g. "$", "€", "£" or, for codes without a common glyph,
     * the code itself followed by a space ("EGP ", "SAR ", "AED ").
     */
    fun symbolFor(code: String): String = when (normalize(code)) {
        "USD" -> "$"
        "EUR" -> "€"
        "GBP" -> "£"
        "EGP" -> "EGP "
        "SAR" -> "SAR "
        "AED" -> "AED "
        else -> "${normalize(code)} "
    }

    /**
     * Rounds + groups [amount] for display. EGP/SAR/AED (whole-number magnitudes) show no decimals;
     * the symbol-prefixed western currencies (USD/EUR/GBP) keep two so small converted values like
     * "$24.36" stay legible. Always formatted with US grouping so digits read the same in both
     * locales (Arabic numerals are handled by the platform per-locale, but grouping stays stable).
     */
    private fun formatAmount(amount: Double, code: String): String {
        return when (normalize(code)) {
            "EGP", "SAR", "AED" -> String.format(java.util.Locale.US, "%,d", Math.round(amount))
            else -> String.format(java.util.Locale.US, "%,.2f", amount)
        }
    }

    /** Upper-cases + trims a code; blank/unknown falls back to [BASE]. */
    private fun normalize(code: String?): String {
        val c = code?.trim()?.uppercase().orEmpty()
        return if (c.isBlank()) BASE else c
    }

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
}

File: mobile/ios/Sources/CurrencyManager.swift
import Foundation
import Combine

/// App-wide multi-currency display controller.
///
/// All backend money is denominated in EGP. This manager converts an EGP amount
/// into the user's chosen display currency (EGP base + USD/EUR/GBP/SAR/AED) using
/// FX rates fetched once at launch from `GET /api/local/currencies` (falling back
/// to baked-in static rates if that call fails). Conversion is **display-only** —
/// bookings are always created and charged in EGP.
///
/// Injected at the app root via `.environmentObject`, mirroring how
/// `LocalizationManager` is provided, so views observe the same source and the
/// whole tree re-renders the instant the user switches currency or the rates
/// finish loading. The chosen currency persists in `UserDefaults` under
/// `qk_currency` (default "EGP").
@MainActor
final class CurrencyManager: ObservableObject {
    static let shared = CurrencyManager()

    /// `UserDefaults` key the chosen display currency persists under.
    static let storageKey = "qk_currency"

    /// The currency prices are displayed in. Persisted on every change.
    @Published var currency: DisplayCurrency {
        didSet {
            UserDefaults.standard.set(currency.rawValue, forKey: Self.storageKey)
        }
    }

    /// The active FX rates. Starts with the baked-in static table so prices
    /// convert correctly even before the network fetch resolves; replaced by the
    /// fetched rates on success.
    @Published private(set) var rates: CurrencyRates = .fallback

    init() {
        if let saved = UserDefaults.standard.string(forKey: Self.storageKey),
           let parsed = DisplayCurrency(rawValue: saved) {
            currency = parsed
        } else {
            currency = .egp
        }
    }

    /// Fetch the live FX rates from the backend, replacing the baked-in defaults
    /// on success. Fails silently — the static `CurrencyRates.fallback` stays in
    /// place so the switcher still works offline. Safe to call repeatedly (e.g.
    /// at app launch).
    func refreshRates() async {
        guard let fetched = try? await HostService.shared.fetchCurrencyRates() else { return }
        // Keep whatever we have if the payload somehow arrived empty.
        guard !fetched.rates.isEmpty else { return }
        rates = fetched
    }

    /// Convert an EGP amount into the chosen currency's value: `egp * rate`.
    /// Falls back to the base amount when the chosen currency has no rate.
    func convert(_ egp: Double) -> Double {
        let rate = rates.rates[currency.code] ?? (currency == .egp ? 1 : 1)
        return egp * rate
    }

    /// Convert + format an EGP amount in the chosen currency, e.g. "EGP 1,100",
    /// "$22.33", "€20.68". Used for the most visible prices across the app
    /// (listing card, listing detail, reserve/receipt totals).
    func format(_ egp: Double) -> String {
        let value = convert(egp)
        let number = NSNumber(value: value)
        let formatted = Self.numberFormatter(for: currency).string(from: number)
            ?? String(format: "%.\(currency.fractionDigits)f", value)
        return "\(currency.symbol)\(formatted)"
    }

    /// Grouped number formatter (thousands separators) for a currency's fraction
    /// digits. Cached per fraction-digit count so we don't rebuild it per row.
    private static func numberFormatter(for currency: DisplayCurrency) -> NumberFormatter {
        if let cached = formatterCache[currency.fractionDigits] { return cached }
        let f = NumberFormatter()
        f.numberStyle = .decimal
        f.locale = Locale(identifier: "en_US_POSIX")
        f.minimumFractionDigits = currency.fractionDigits
        f.maximumFractionDigits = currency.fractionDigits
        f.usesGroupingSeparator = true
        formatterCache[currency.fractionDigits] = f
        return f
    }

    private static var formatterCache: [Int: NumberFormatter] = [:]
}
